File: planner.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Set, Tuple

from models import CraftStep, PlanStep, Recipe, SupplyStep
from util.itemkey import normalize
from util.log import debug, info


@dataclass
class PlanContext:
    resource: object
    recipes: Dict[str, List[Recipe]]
    reachable: Set[str] = field(default_factory=set)
    virtual_stock: Dict[str, int] = field(default_factory=dict)
    initial_stock: Dict[str, int] = field(default_factory=dict)
    stock_remaining: Dict[str, int] = field(default_factory=dict)
    stack: Set[str] = field(default_factory=set)
    plan: List[PlanStep] = field(default_factory=list)


def select_recipe(recipes: List[Recipe]) -> Recipe:
    return sorted(recipes, key=lambda r: r.priority, reverse=True)[0]


def get_virtual_stock(ctx: PlanContext, item: str) -> int:
    if item not in ctx.virtual_stock:
        raise RuntimeError("item_not_initialized")
    return int(ctx.virtual_stock[item])


def compute_reachable(target_item_key: str, recipes_by_output: Dict[str, List[Recipe]]) -> Set[str]:
    reachable: Set[str] = set()

    def dfs_item(item_key: str) -> None:
        key = normalize(item_key)
        if key in reachable:
            return
        reachable.add(key)
        for recipe in recipes_by_output.get(key, []):
            for stack in recipe.inputs:
                dfs_item(stack.item)

    dfs_item(target_item_key)
    return reachable


def initialize_stock(ctx: PlanContext, reachable_items: Set[str]) -> None:
    for item in reachable_items:
        value = int(ctx.resource.get(item))
        ctx.initial_stock[item] = value
        ctx.virtual_stock[item] = value
        ctx.stock_remaining[item] = value


def get_output_for_item(recipe: Recipe, item: str):
    for output in recipe.outputs:
        if normalize(output.item) == item:
            return output
    return None


def plan_need(ctx: PlanContext, item: str, count: int) -> Tuple[bool, str | None]:
    item = normalize(item)
    debug(f"plan_need start item={item} count={count}")
    if item in ctx.stack:
        return False, "cycle"
    if item not in ctx.reachable:
        return False, "item_not_reachable"

    try:
        available = get_virtual_stock(ctx, item)
    except RuntimeError:
        return False, "item_not_initialized"
    use = min(count, max(0, available))
    if use > 0:
        from_stock = min(use, ctx.stock_remaining.get(item, 0))
        if from_stock > 0:
            ctx.stock_remaining[item] -= from_stock
            if ctx.stock_remaining[item] < 0:
                return False, "supply_exceeds_stock"
            ctx.plan.append(SupplyStep(item=item, count=from_stock))
            debug(f"plan_need supply item={item} count={from_stock}")
        ctx.virtual_stock[item] = available - use
        if ctx.virtual_stock[item] < 0:
            return False, "virtual_stock_negative"
        debug(f"plan_need consume item={item} use={use} stock_before={available}")

    remain = count - use

    if remain == 0:
        debug(f"plan_need satisfied item={item}")
        return True, None

    recipes = ctx.recipes.get(item)
    if not recipes:
        debug(f"plan_need no recipe for item={item} remain={remain}")
        return False, "no_recipe_or_stock"

    ctx.stack.add(item)
    try:
        recipe = select_recipe(recipes)
        debug(f"plan_need selected recipe={recipe.id} for item={item}")
        out = get_output_for_item(recipe, item)
        if out is None or out.count <= 0:
            return False, "no_matching_output"
        times = (remain + out.count - 1) // out.count
        debug(f"plan_need recipe={recipe.id} times={times} out_count={out.count} remain={remain}")

        for input_item in recipe.inputs:
            debug(f"plan_need need input item={input_item.item} count={input_item.count * times}")
            ok, err = plan_need(ctx, input_item.item, input_item.count * times)
            if not ok:
                return False, err

        ctx.plan.append(CraftStep(recipe=recipe, times=times))

        for output in recipe.outputs:
            out_item = normalize(output.item)
            produced_count = output.count * times
            current = ctx.virtual_stock.get(out_item, 0)
            ctx.virtual_stock[out_item] = current + produced_count
            debug(f"plan_need produced item={output.item} count={produced_count}")

        current_item = get_virtual_stock(ctx, item)
        if __debug__:
            assert current_item - remain >= 0
        ctx.virtual_stock[item] = current_item - remain
        debug(f"plan_need consume produced item={item} count={remain}")
    finally:
        ctx.stack.discard(item)

    return True, None


def compress_supplies(plan: List[PlanStep]) -> List[PlanStep]:
    totals: Dict[str, int] = {}
    order: List[str] = []
    crafts: List[PlanStep] = []

    for step in plan:
        if isinstance(step, SupplyStep):
            if step.item not in totals:
                order.append(step.item)
                totals[step.item] = 0
            totals[step.item] += step.count
        else:
            crafts.append(step)

    debug(f"compress_supplies merged_items={len(order)}")
    merged_supplies = [
        SupplyStep(item=item, count=totals[item])
        for item in order
        if totals[item] > 0
    ]
    return merged_supplies + crafts


def compress_crafts(plan: List[PlanStep]) -> List[PlanStep]:
    merged: List[PlanStep] = []
    merged_count = 0
    for step in plan:
        if not isinstance(step, CraftStep):
            merged.append(step)
            continue
        last = merged[-1] if merged else None
        if (
            last
            and isinstance(last, CraftStep)
            and last.recipe.id == step.recipe.id
            and last.recipe.machine == step.recipe.machine
        ):
            merged[-1] = CraftStep(recipe=last.recipe, times=last.times + step.times)
            merged_count += 1
            continue
        merged.append(step)
    debug(f"compress_crafts merged_steps={merged_count}")
    return merged


def plan(target_item_key: str, target_count: int, recipes_by_output: Dict[str, List[Recipe]], resource_provider: object) -> Tuple[bool, List[PlanStep] | str]:
    reachable = compute_reachable(target_item_key, recipes_by_output)
    ctx = PlanContext(resource=resource_provider, recipes=recipes_by_output, reachable=reachable)
    debug(f"plan start target={target_item_key} count={target_count}")
    if hasattr(resource_provider, "prepare"):
        resource_provider.prepare(reachable)
    initialize_stock(ctx, reachable)
    if hasattr(resource_provider, "snapshot"):
        resource_provider.snapshot()
    ok, err = plan_need(ctx, target_item_key, target_count)
    if not ok:
        return False, err or "plan_failed"

    if __debug__:
        target_item = normalize(target_item_key)
        assert ctx.virtual_stock.get(target_item, 0) >= 0
        for value in ctx.virtual_stock.values():
            assert value >= 0
        for value in ctx.stock_remaining.values():
            assert value >= 0

    ctx.plan = compress_supplies(ctx.plan)
    ctx.plan = compress_crafts(ctx.plan)

    for step in ctx.plan:
        if isinstance(step, SupplyStep):
            info(f"PLAN supply {step.item} x{step.count}")
        elif isinstance(step, CraftStep):
            info(f"PLAN craft {step.recipe.id} x{step.times}")

    return True, ctx.plan

File: lua/core/planner.lua
local util = require("core.util")
local steps = require("core.steps")

local M = {}

local function compute_reachable(target, recipes_by_output)
  local reachable = {}
  local function dfs(item_key)
    local key = util.normalize(item_key)
    if reachable[key] then return end
    reachable[key] = true
    local recipes = recipes_by_output[key] or {}
    for _, recipe in ipairs(recipes) do
      for _, input in ipairs(recipe.inputs) do
        dfs(input.item)
      end
    end
  end
  dfs(target)
  return reachable
end

local function select_recipe(recipes)
  table.sort(recipes, function(a, b)
    return (a.priority or 0) > (b.priority or 0)
  end)
  return recipes[1]
end

local function get_output_for_item(recipe, item)
  for _, output in ipairs(recipe.outputs) do
    if util.normalize(output.item) == item then
      return output
    end
  end
  return nil
end

local function compress_supplies(plan)
  local totals = {}
  local order = {}
  local crafts = {}
  for _, step in ipairs(plan) do
    if step.kind == "supply" then
      if not totals[step.item] then
        totals[step.item] = 0
        table.insert(order, step.item)
      end
      totals[step.item] = totals[step.item] + step.count
    else
      table.insert(crafts, step)
    end
  end
  local merged = {}
  for _, item in ipairs(order) do
    if totals[item] > 0 then
      table.insert(merged, steps.supply(item, totals[item]))
    end
  end
  for _, step in ipairs(crafts) do
    table.insert(merged, step)
  end
  return merged
end

local function compress_crafts(plan)
  local merged = {}
  local last = nil
  for _, step in ipairs(plan) do
    if step.kind == "craft" then
      if last and last.kind == "craft" and last.recipe.id == step.recipe.id and last.recipe.machine == step.recipe.machine then
        last.times = last.times + step.times
      else
        table.insert(merged, step)
        last = step
      end
    else
      table.insert(merged, step)
      last = step
    end
  end
  return merged
end

local function plan_need(ctx, item, count)
  local key = util.normalize(item)
  if ctx.stack[key] then
    return false, "cycle"
  end
  if not ctx.reachable[key] then
    return false, "item_not_reachable"
  end

  local available = ctx.virtual_stock[key]
  if available == nil then
    return false, "item_not_initialized"
  end

  local use = math.min(count, math.max(0, available))
  if use > 0 then
    local from_stock = math.min(use, ctx.stock_remaining[key] or 0)
    if from_stock > 0 then
      ctx.stock_remaining[key] = ctx.stock_remaining[key] - from_stock
      if ctx.stock_remaining[key] < 0 then
        return false, "supply_exceeds_stock"
      end
      table.insert(ctx.plan, steps.supply(key, from_stock))
    end
    ctx.virtual_stock[key] = available - use
    if ctx.virtual_stock[key] < 0 then
      return false, "virtual_stock_negative"
    end
  end

  local remain = count - use
  if remain == 0 then
    return true
  end

  local recipes = ctx.recipes[key]
  if not recipes or #recipes == 0 then
    return false, "no_recipe_or_stock"
  end

  ctx.stack[key] = true

  local recipe = select_recipe(recipes)
  local out = get_output_for_item(recipe, key)
  if not out or out.count <= 0 then
    ctx.stack[key] = nil
    return false, "no_matching_output"
  end
  local times = math.floor((remain + out.count - 1) / out.count)

  for _, input in ipairs(recipe.inputs) do
    local ok, err = plan_need(ctx, input.item, input.count * times)
    if not ok then
      ctx.stack[key] = nil
      return false, err
    end
  end

  table.insert(ctx.plan, steps.craft(recipe, times))

  for _, output in ipairs(recipe.outputs) do
    local out_item = util.normalize(output.item)
    local produced = output.count * times
    ctx.virtual_stock[out_item] = (ctx.virtual_stock[out_item] or 0) + produced
  end

  local current = ctx.virtual_stock[key] or 0
  if current - remain < 0 then
    ctx.stack[key] = nil
    return false, "virtual_stock_negative"
  end
  ctx.virtual_stock[key] = current - remain
  ctx.stack[key] = nil
  return true
end

function M.plan(target_item_key, target_count, recipes_by_output, resource_provider)
  local reachable = compute_reachable(target_item_key, recipes_by_output)
  local ctx = {
    resource = resource_provider,
    recipes = recipes_by_output,
    reachable = reachable,
    virtual_stock = {},
    stock_remaining = {},
    stack = {},
    plan = {},
  }

  if resource_provider.prepare then
    resource_provider:prepare(reachable)
  end

  for item, _ in pairs(reachable) do
    local value = resource_provider:get(item)
    ctx.virtual_stock[item] = value
    ctx.stock_remaining[item] = value
  end

  if resource_provider.snapshot then
    resource_provider:snapshot()
  end

  local ok, err = plan_need(ctx, target_item_key, target_count)
  if not ok then
    return false, err
  end

  local result = compress_supplies(ctx.plan)
  result = compress_crafts(result)
  return true, result
end

return M
